astropy/units/quantity.py

```diff
@@ -286,6 +317,102 @@ class Quantity(np.ndarray):
+                and all other elements are added metadata (See Annotated).
+
+            Returns
+            -------
+            :class:`~astropy.units.typing.Annotated` instance
+                The origin type is a `~typing.TypeVar` with name
+                ``repr(cls)` and upper bound  ``cls``.
+                The metadata is ordered by [unit, other, ...]
+
+                Schematically,
+                ``Annotated[TypeVar(repr(cls), bound=cls), *unit_annotation]``
+
+                As a classmethod, the type is the class, ie ``Quantity``
+                produces an ``Annotated[Quantity, ...]`` while a subclass
+                like ``Length`` returns ``Annotated[Length, ...]``.
+
+            Raises
+            ------
+            NotImplementedError
+                If (0th elt of) unit_annotation is not a |Unit| or |Quantity|
+                or str
+            ValueError
+                If (0th elt of) unit_annotation is str but not a physical type.
+
+            Examples
+            --------
+            Create a unit-aware Quantity type annotation
+
+                >>> Quantity[Unit("s")]
+                Annotated[~<class 'Quantity'>, Unit("s")]
+
+            This is equivalent, but easier than, manually creating the
+            annotation. In Python 3.9+ ``Annotated`` is built into
+            :mod:`~typing`. For older python versions, the optional dependency
+            ``typing_extensions`` provides the same class. When neither are
+            available we default to a run-time only compatibility class (in
+            ``astropy.units._typing``) with reduced functionality. The best
+            implementation is automatically detected and used.
+
+                >>> from astropy.units._typing import Annotated  # imports best
+                >>> Annotated[Quantity, Unit("s")]
+                Annotated[~<class 'Quantity'>, Unit("s")]
+
+            With Python 3.9+ or ``typing_extensions`` Quantity types
+            can be used as static type annotations.
+
+                >>> def func(x: Quantity[u.kpc]) -> Quantity[u.m]:
+                ...     return x << u.m
+
+            """
+            if isinstance(unit_annotation, tuple):  # unit and annotations
+                unit = unit_annotation[0]
+                annotations = unit_annotation[1:]
+            else:  # just unit
+                unit = unit_annotation
+                annotations = ()
+
+            # The allowed unit types
+            if isinstance(unit, (UnitBase, Quantity, str)):
+                # check unit / physical type is allowed
+                # Note: converts Quantity to a Unit(Quantity)
+                unit = _parse_allowed_type_hint(unit, detailed_exception=True)
+
+                # Creating a TypeVar with bound `cls` ensures that Quantity
+                # (or subclass) can be used by a static type checker.
+                origin = _quantitytype_registry.setdefault(
```

Comment:
I'm wondering about the registry: what is the advantage to having it? It doesn't seem actually used.

Reply:
``dict.setdefault`` is actually a compact ``getter`` & ``setter`` method. If it's present in the dict, then it returns it, else it sets from the default value passed and returns that.

Reply:
As far as I can tell, every call to `typing.TypeVar(x)` is an unique instance. In order to do equality comparisons it seems necessary to cache the TypeVar.

Reply:
The unique instance seems important, but then I would not use `setdefault`, since you are creating new instances only to discard them immediately. But looking closer, is it even possible for the dictionary not to have the entry? How can one define the class and not run through `__subclass_init__`?

Reply:
🤔 . It does only need one or the other. I'm removing the `setdefault`.